ext/src/ruby_api/wasi_config.rs

```diff
@@ -0,0 +1,283 @@
+use super::root;
+use crate::{error, helpers::WrappedStruct};
+use magnus::{
+    function, gc, method, DataTypeFunctions, Error, Module, Object, RArray, RHash, RString,
+    TypedData,
+};
+use std::cell::RefCell;
+use std::{fs::File, path::PathBuf};
+use wasi_common::pipe::ReadPipe;
+use wasmtime_wasi::{WasiCtx, WasiCtxBuilder};
+
+enum ReadStream {
+    Inherit,
+    Path(RString),
+    String(RString),
+}
+
+impl ReadStream {
+    pub fn mark(&self) {
+        match self {
+            Self::Inherit => (),
+            Self::Path(s) => gc::mark(*s),
+            Self::String(s) => gc::mark(*s),
+        }
+    }
+}
+
+enum WriteStream {
+    Inherit,
+    Path(RString),
+}
+impl WriteStream {
+    pub fn mark(&self) {
+        match self {
+            Self::Inherit => (),
+            Self::Path(v) => gc::mark(*v),
+        }
+    }
+}
+
+#[derive(Default)]
+struct WasiConfigInner {
+    stdin: Option<ReadStream>,
+    stdout: Option<WriteStream>,
+    stderr: Option<WriteStream>,
+    env: Option<RHash>,
+    args: Option<RArray>,
+}
+
+macro_rules! mark_some {
+    ($expr:expr) => {{
+        if let Some(v) = $expr.as_ref() {
+            v.mark();
+        }
+    }};
+}
+
+impl WasiConfigInner {
+    pub fn mark(&self) {
+        mark_some!(self.stdin);
+        mark_some!(self.stdout);
+        mark_some!(self.stderr);
+        if let Some(v) = self.env.as_ref() {
+            gc::mark(*v);
+        }
+        if let Some(v) = self.args.as_ref() {
+            gc::mark(*v);
+        }
+    }
+}
+
+/// @yard
+/// WASI configuration builder to be used in {Store#configure_wasi}.
+///
+/// Instance methods mutate the current object and return +self+.
+///
+/// @see https://docs.rs/wasmtime-wasi/latest/wasmtime_wasi/sync/struct.WasiCtxBuilder.html
+///   Wasmtime's doc for WasiCtxBuilder.
+// #[derive(Debug)]
+#[derive(Default, TypedData)]
+#[magnus(class = "Wasmtime::WasiConfig", size, mark, free_immediatly)]
+pub struct WasiConfig {
+    inner: RefCell<WasiConfigInner>,
+}
+
+impl DataTypeFunctions for WasiConfig {
+    fn mark(&self) {
+        self.inner.borrow().mark();
+    }
+}
+
+type RbSelf = WrappedStruct<WasiConfig>;
+
+impl WasiConfig {
```

Comment:
The other question that comes to mind here, is why did we decided to go with a builder patter rather than having plain setters?

Reply:
> What's the reason to deviate from `WasiCtx`?

As I alluded to in another comment, this `WasiConfig` is more like the C API's `wasi_config_t`. It lets you describe a WASI context once, and we'll create the `WasiCtx` for the config when needed. That's why I stuck to the "wasi config" name.

The Rust `WasiCtx` is what you get after "applying" (for lack of a better word) the Ruby `WasiConfig` to a Ruby `Store`. `WasiCtx` is already configured with random, clock, etc., the `WasiConfig` is not. We could eventually add methods to `WasiConfig` so it can build a deterministic wasi context, or maybe even a `WasiCtx` that's backed by Ruby Procs (although that's possibly slower).

Reply:
> why did we decided to go with a builder patter rather than having plain setters?

I saw that `WasiCtxBuilder` uses a builder pattern and thought that was pretty elegant. It can remove the need to repeat a variable name over and over.

Also, when I think of setters, I think of simple property assigns. In this case, some settings are mutually exclusive (as expressed with the enums), thus setting a field can have effect on other fields. The current builder API does not make this obvious, but it still feels better than setters. We could make this clearer, at the cost of extra boilerplate, making the API more complex, and extra object allocations (off the hot path):

```ruby
# With extra object:
Wastime::WasiConfig.new
  .stdout(Wastime::WasiConfig::File.new("tmp/stdout"))
  .stdin(Wasmtime::WasiConfig::String.new("stdin bytes"))

# With blocks:
Wastime::WasiConfig.new
  .stdout { |stdout| stdout.file("tmp/stdout") }
  .stdin { |stdin| stdin.string("stdin bytes") }
```

I like these APIs, but they feel too complex for the task at hand.

Reply:
> The Rust WasiCtx is what you get after "applying" (for lack of a better word) the Ruby WasiConfig 

I should've been more clear here. I meant why we deviated from the `WasiCtx` and `WasiCtxBuilder` naming. Your explanation above makes sense, but I'm not sure that the C-API should be treated as "canonical", instead I think that our canonical should be the Rust one where possible. That's why I think that using `WasiCtx` should be preferred. 

I'm good with the setters part, I was asking more out of curiosity.